### ISSUE TITLE
[oneDNN] Fused SDPA GPU page-fault / device lost when head_dim=64 with a rank-4 attn_mask spanning more than one tile on both axes (Sq>=17, Sk>=129); rank-2 mask is safe

### ISSUE THREAD

**Issue by 18856672421** [oneapi-src/oneDNN#5933]
## Summary

On an Intel Arc (Alchemist-class) iGPU, the oneDNN fused SDPA kernel used by
`F.scaled_dot_product_attention` hard-crashes the device (GPU page fault →
device lost / process abort, **not** a catchable Python exception) **only when
all four** of the following hold:

1. `attn_mask` is a **rank-4** mask, shape `[B, 1, Sq, Sk]` (broadcast over heads);
2. `head_dim == 64`;
3. `Sq >= 17` (i.e. query axis spans a 2nd 16-wide tile);
4. `Sk >= 129` (i.e. key axis spans a 2nd 128-wide tile).

Changing **any single** condition makes it safe:

- `head_dim != 64` (e.g. 32 / 96 / 128) → OK;
- `Sq <= 16` **or** `Sk <= 128` (single tile on either axis) → OK;
- passing the **same mask reshaped to rank-2** `[Sq, Sk]` → OK;
- `attn_mask=None` → OK.

The crash is independent of batch size and of GQA: it reproduces at `B=1` and
with equal q/k/v head counts (no `enable_gqa`). It looks like an out-of-bounds
access in the fused kernel's second-tile handling, specific to head_dim=64,
that only triggers when a rank-4 mask is present.

This is related to (but distinct from) the known Alchemist head_dim=64 SDPA
problem in uxlfoundation/oneDNN#5510, which reports *wrong results* for d=64 on
Arc 770 / Arc 140T; here the failure mode is a hard device crash gated on the
rank-4 mask and the 16/128 tile boundaries.

## Environment

- PyTorch: **2.13.0+xpu** (official XPU wheel, Windows)
- Intel GPU driver: **32.0.101.8991**
- GPU: **Intel Arc Graphics 140T** (Alchemist / Xe-HPG class)
- OS: Windows 11
- oneDNN / MKL-DNN: 3.12.0
- dtype: `torch.bfloat16`
- Backend note on this build:
  - FLASH_ATTENTION: unavailable ("Torch XPU was not compiled with flash attention");
  - EFFICIENT_ATTENTION: prints "Memory Efficient Attention backend is not
    supported on XPU, falling back to math backend" → effectively MATH;
  - so the only fused path is the oneDNN fused kernel, which is the one that crashes.

## Minimal reproducer (crashes the process)

```python
import torch
import torch.nn.functional as F

dev, dt = "xpu", torch.bfloat16
B, H, Sq, Sk, D = 1, 14, 17, 129, 64          # <-- all four conditions met

q = torch.randn(B, H, Sq, D, device=dev, dtype=dt)
k = torch.randn(B, H, Sk, D, device=dev, dtype=dt)
v = torch.randn(B, H, Sk, D, device=dev, dtype=dt)
m = torch.ones(B, 1, Sq, Sk, device=dev, dtype=torch.bool)   # rank-4 broadcast mask

out = F.scaled_dot_product_attention(q, k, v, attn_mask=m)   # GPU page fault / device lost
torch.xpu.synchronize()
```

### Single-change controls (all OK — proves the four-way conjunction)

```python
# (a) different head_dim
F.scaled_dot_product_attention(q[..., :32], k[..., :32], v[..., :32], attn_mask=m)  # D=32  OK
# (b) query within one 16-tile
F.scaled_dot_product_attention(q[:, :, :16],  k, v, attn_mask=m[:, :, :16])         # Sq=16 OK
# (c) key within one 128-tile
F.scaled_dot_product_attention(q, k[:, :, :128], v[:, :, :128], attn_mask=m[..., :128])  # Sk=128 OK
# (d) identical mask, reshaped to rank-2
F.scaled_dot_product_attention(q, k, v, attn_mask=m.reshape(Sq, Sk))                # rank2 OK
# (e) no mask
F.scaled_dot_product_attention(q, k, v)                                             # OK
```

## Boundary sweep (run each config in a subprocess — a crash kills the worker)

```python
# probe.py  —  python probe.py            (driver, summarizes)
#             python probe.py w <D> <Sq> <Sk> <mask_rank>   (one isolated worker)
import sys, subprocess

def worker(D, Sq, Sk, mask_rank):
    import torch
    import torch.nn.functional as F
    dev, dt = "xpu", torch.bfloat16
    B, H = 1, 14
    q = torch.randn(B, H, Sq, D, device=dev, dtype=dt)
    k = torch.randn(B, H, Sk, D, device=dev, dtype=dt)
    v = torch.randn(B, H, Sk, D, device=dev, dtype=dt)
    if mask_rank == 4:
        m = torch.ones(B, 1, Sq, Sk, device=dev, dtype=torch.bool)
    else:
        m = torch.ones(Sq, Sk, device=dev, dtype=torch.bool)
    out = F.scaled_dot_product_attention(q, k, v, attn_mask=m)
    torch.xpu.synchronize()
    print("OK", tuple(out.shape), flush=True)

def probe(D, Sq, Sk, mask_rank):
    r = subprocess.run(
        [sys.executable, __file__, "w", str(D), str(Sq), str(Sk), str(mask_rank)],
        capture_output=True, text=True, timeout=120)
    low = (r.stdout + r.stderr).lower()
    if "ok" in low:
        return "OK"
    if any(s in low for s in ("device lost", "segmentation fault", "page fault",
                              "abort", "fault", "crash")):
        return "CRASH"
    return "ERR: " + (r.stderr.strip().splitlines() or ["?"])[-1][:80]

def main():
    grid = [
        # (D,  Sq,   Sk,   mask_rank)
        (64,  16,  128, 4),   # single tile both axes
        (64,  17,  128, 4),   # 2nd Q tile only
        (64,  16,  129, 4),   # 2nd K tile only
        (64,  17,  129, 4),   # 2nd tile on BOTH axes  -> crash
        (64,  33,  256, 4),   # further over tiles      -> crash
        (32,  17,  129, 4),   # same over-tile, hd!=64  -> OK
        (128, 17,  129, 4),   # same over-tile, hd!=64  -> OK
        (64,  17,  129, 2),   # same shape, rank-2 mask -> OK
        (64, 130, 1024, 4),   # realistic long context  -> crash
    ]
    for D, Sq, Sk, mr in grid:
        print(f"hd={D:<3} Sq={Sq:<4} Sk={Sk:<4} mask_rank={mr} -> {probe(D, Sq, Sk, mr)}",
              flush=True)

if __name__ == "__main__":
    if len(sys.argv) >= 6 and sys.argv[1] == "w":
        worker(int(sys.argv[2]), int(sys.argv[3]), int(sys.argv[4]), int(sys.argv[5]))
    else:
        main()
```

### Observed result

hd=64  Sq=16   Sk=128  mask_rank=4 -> OK
hd=64  Sq=17   Sk=128  mask_rank=4 -> OK
hd=64  Sq=16   Sk=129  mask_rank=4 -> OK
hd=64  Sq=17   Sk=129  mask_rank=4 -> ERR: level_zero backend failed with error: 20 (UR_RESULT_ERROR_DEVICE_LOST)
hd=64  Sq=33   Sk=256  mask_rank=4 -> ERR: level_zero backend failed with error: 20 (UR_RESULT_ERROR_DEVICE_LOST)
hd=32  Sq=17   Sk=129  mask_rank=4 -> OK
hd=128 Sq=17   Sk=129  mask_rank=4 -> OK
hd=64  Sq=17   Sk=129  mask_rank=2 -> OK
hd=64  Sq=130  Sk=1024 mask_rank=4 -> ERR: level_zero backend failed with error: 20 (UR_RESULT_ERROR_DEVICE_LOST)

=> crash requires **head_dim=64 AND rank-4 mask AND crossing both the 16-Q and
128-K tile boundaries**. Batch size (reproduces at B=1) and GQA (reproduces with
equal q/k/v heads) are not required.

## Real-world impact

Hit while running Qwen2-VL inference (transformers SDPA attention interface) via
MinerU on this GPU. In the actual model q has 14 heads vs 2 KV heads (GQA,
`enable_gqa=True`) and the padding mask is rank-4 `[B,1,Sq,Sk]`; once the
prefill context grows past the thresholds above (e.g. Sq=685, Sk=685, hd=64),
every attention layer crashes and takes down the whole process.

## Workaround (user-side, confirms the diagnosis)

Force the MATH backend

This is numerically correct and prevents the device crash, and it needs no
reshaping or per-batch splitting. However, MATH is a non-fused implementation
and is substantially slower than the native oneDNN fused kernel on this iGPU,
which noticeably slows down every affected attention layer during prefill.
That performance gap — having to stay on the slow MATH path for a shape the
fused kernel ought to support — is the reason for this report.

## Expected behavior

The fused oneDNN SDPA kernel should compute correct results for
head_dim=64 with a rank-4 mask that spans multiple 16/128 tiles (matching
CUDA and the rank-2-mask / other-head-dim behavior)

**Comment by TaoLv** [comment]
Thank you for the report, @18856672421 .  Could you please run the problematic case with environment variable `ONEDNN_GRAPH_DUMP=subgraph` and share the JSON file so I can reproduce the issue with it? Thanks.

**Comment by 18856672421** [comment]
Hi, @TaoLv 
Thanks! Attached is the fused SDPA subgraph from the actual failing run
(ONEDNN_GRAPH_DUMP=subgraph, torch 2.13.0+xpu, Arc 140T, bf16). It is the
external-mask path (not the internal is_causal one): GQA 5D q/k/v
[7,2,7,454,64] (14 q-heads / 2 kv-heads, head_dim=64) with an external
additive bf16 mask [7,1,1,454,454] fed through mask_add (Add); Sq=Sk=454, so
both the 16-Q and 128-K tile boundaries are crossed. The device is lost
right after this partition is executed. I can also provide the bool-mask
minimal-repro dump if useful.

[graph-104658-5349023115352890067.json](https://github.com/user-attachments/files/31615280/graph-104658-5349023115352890067.json)

**Comment by TaoLv** [comment]
Hi @18856672421 , I see the JSON file can work properly with the latest oneDNN release (v3.13.2, see the log below). You may want to switch to the latest version of oneDNN and try your workload again, or wait for PyTorch 2.15 which will include oneDNN v3.13.2.

```
E:\lvtao\oneDNN\build>tests\benchdnn\benchdnn.exe --graph --engine=gpu --case=E:\lvtao\graph-104658-5349023115352890067.json
onednn_verbose,v1,info,oneDNN v3.13.2 (commit 71094df451909e95ceeefc8f64c275cdcbbcdff5)
onednn_verbose,v1,info,cpu,runtime:OpenMP,nthr:20
onednn_verbose,v1,info,cpu,isa:Intel AVX2
onednn_verbose,v1,info,gpu,runtime:SYCL
onednn_verbose,v1,info,gpu,engine,sycl gpu device count:1
onednn_verbose,v1,info,gpu,engine,0,backend:OpenCL,name:Intel(R) Arc(TM) A770 Graphics,driver_version:32.0.101,binary_kernels:enabled
onednn_verbose,v1,info,graph,backend,0:dnnl_backend
onednn_verbose,v1,primitive,info,template:operation,engine,primitive,implementation,prop_kind,memory_descriptors,attributes,auxiliary,problem_desc,exec_time
onednn_verbose,v1,graph,info,template:operation,engine,partition_id,partition_kind,op_names,data_formats,logical_tensors,fpmath_mode,implementation,backend,exec_time
onednn_verbose,v1,common,info,filter format is enabled, hit components: graph
onednn_verbose,v1,primitive,exec,,0.341354
onednn_verbose,v1,primitive,exec,,1.48583
onednn_verbose,v1,primitive,exec,,0.732395
onednn_verbose,v1,primitive,exec,,0.716666
onednn_verbose,v1,primitive,exec,,0.292916
onednn_verbose,v1,graph,exec,gpu,100002,sdp,matmul_qk;scale_mul;mask_add;softmax;matmul_v,,in0_bf16:0:strided:undef:7x2x7x454x64:406784s448s64s896s1 in1_bf16:1:strided:undef:7x2x1x454x64:58112s29056s29056s64s1 in2_f32:2:strided:host_scalar:0:2 in3_bf16:4:strided:undef:7x1x1x454x454:206116s206116s206116s454s1 in4_bf16:5:strided:undef:7x2x1x454x64:58112s29056s29056s64s1 out0_bf16:6:strided:undef:7x2x7x454x64:406784s448s64s896s1,fpm:strict,scratchpad:library:0,sdp_primitive_kernel_t,dnnl_backend,1.4637
onednn_verbose,v1,primitive,exec,gpu,sdpa,ocl:micro:reusable,forward_inference,query:bf16::blocked:acbd::f0 key:bf16::blocked:abdc::f0 val:bf16::blocked:abcd::f0 msk:bf16::blocked:abcd::f0 dst:bf16::blocked:acbd::f0,attr-scratchpad:user,alg:softmax_accurate_inf_as_zero msk:2d scl:mul:f32:host,7x14x454x64:7x2x64x454:7x2x454x64,0.490312
0:PASSED (2102 ms) __REPRO: --graph --engine=gpu --case=E:\lvtao\graph-104658-5349023115352890067.json
tests:1 passed:1 skipped:0 mistrusted:0 unimplemented:0 invalid_arguments:0 failed:0 listed:0
total: 2.11s; create_pd: 0.37s (18%); create_prim: 0.03s (1%); fill: 0.00s (0%); execute: 0.18s (9%); compute_ref: 0.00s (0%); compare: 0.00s (0%);

E:\lvtao\oneDNN\build>
```

**Comment by 18856672421** [comment]
@TaoLv 
Thanks for verifying with oneDNN v3.13.2 on the A770 — glad it PASSED. That
confirms the fused SDPA mask path was fixed upstream. I'm on torch
2.13.0+xpu (bundled oneDNN 3.12.x) on an Arc 140T, so I'll wait for
PyTorch 2.15 (oneDNN 3.13.2) and re-test the workload; meanwhile I'll keep
the MATH fallback. Happy to close once I can confirm on 2.15.
One more question if possible: is there a scheduled timeline for when a
PyTorch build bundling oneDNN 3.13.2 will be available? As of today
(2026-08-31) the latest XPU nightly (torch-2.15.0.dev20260830) still bundles
oneDNN 3.12.3 (verified from the torch_xpu.dll inside the wheel), and I
couldn't find a 3.13.x submodule bump PR in pytorch/pytorch yet — the only
open oneDNN upgrade seems to be the draft #194400 (3.12.4). Should I watch
for a specific nightly date, or wait for the 2.15 stable release? Thanks!

**Comment by TaoLv** [comment]
@Lufinch is preparing to bump oneDNN version for PyTorch main and 2.15.

**Comment by 18856672421** [comment]
@TaoLv ,Thanks for the heads-up. I'll watch for LuFinch's oneDNN bump landing on main/2.15 and re-run both my boundary repro and the real Qwen2-VL workload on the Arc 140T once it's in a nightly, then report back here. Happy to help verify on 140T if that's useful.
